src/app/interaction-model/Model.h

```diff
@@ -0,0 +1,123 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <lib/core/TLVReader.h>
+#include <lib/core/TLVWriter.h>
+
+#include <app/AttributeValueDecoder.h>
+#include <app/AttributeValueEncoder.h>
+
+#include <app/interaction-model/Actions.h>
+#include <app/interaction-model/InvokeResponder.h>
+#include <app/interaction-model/IterationTypes.h>
+#include <app/interaction-model/OperationTypes.h>
+
+namespace chip {
+namespace app {
+namespace InteractionModel {
+
+/// Represents operations against a matter-defined data model.
+///
+/// Class is SINGLE-THREADED:
+///   - operations are assumed to only be ever run in a single event-loop
+///     thread or equivalent
+///   - class is allowed to attempt to cache indexes/locations for faster
+///     lookups of things (e.g during iterations)
+class Model : public AttributeTreeIterator
+{
+public:
+    virtual ~Model() = default;
+
+    // `actions` pointers  will be guaranteed valid until Shutdown is called()
+    virtual CHIP_ERROR Startup(InteractionModelActions actions)
+    {
+        mActions = actions;
+        return CHIP_NO_ERROR;
+    }
+    virtual CHIP_ERROR Shutdown() = 0;
+
+    // During the transition phase, we expect a large subset of code to require access to
+    // event emitting, path marking and other operations
+    virtual InteractionModelActions CurrentActions() { return mActions; }
+
+    /// List reading has specific handling logic:
+    ///   `state` contains in/out data about the current list reading. MUST start with kInvalidListIndex on first call
+    ///
+    /// Return codes:
+    ///   CHIP_ERROR_MORE_LIST_DATA_AVAILABLE (NOTE: new error defined for this purpose)
```

Comment:
I will revise this as I implement ReadAttribute. I assumed I may touch things around when implementing, so I did not define this yet.

Reply:
Then please make it clear that this is an error _to be defined_.  Because what the text says is that is has been defined, but it hasn't, and that really confused me.

Reply:
I would like to do these updates when these functions get implemented and have a final validation pass on the API. It is not practical to try to fully design the API as even as it stands reviews took an extremely long time.